File: ios/Sync/TokenStore.swift
import Foundation

/// Persists the two bulk-upload auth credentials in a plain `UserDefaults` suite namespaced to
/// this fork: the `x-authorization` bearer token and the `x-sso-token`.
///
/// Deliberately NOT the Keychain: the sync layer must read these natively while JS is dead
/// (background / terminated uploads), and the integrating app already keeps the same tokens in
/// plain storage (AsyncStorage). Plain `UserDefaults` is native-readable with no Keychain
/// dependency. The app pushes tokens in via `setAuthToken`/`setSsoToken`; nothing is read back
/// from JS at upload time.
@objc public class TokenStore: NSObject {

  @objc public static let shared = TokenStore()

  private let defaults: UserDefaults

  private override init() {
    self.defaults = UserDefaults(suiteName: TokenStore.suiteName) ?? .standard
    super.init()
  }

  /// The `x-authorization` bearer token (sent as `Bearer <token>`).
  var authToken: String? {
    get { defaults.string(forKey: Keys.auth) }
    set { write(Keys.auth, newValue) }
  }

  /// The `x-sso-token` value (raw JWT, sent without a `Bearer` prefix).
  var ssoToken: String? {
    get { defaults.string(forKey: Keys.sso) }
    set { write(Keys.sso, newValue) }
  }

  private func write(_ key: String, _ value: String?) {
    if let value = value, !value.isEmpty {
      defaults.set(value, forKey: key)
    } else {
      defaults.removeObject(forKey: key)
    }
  }

  private static let suiteName = "com.backgroundlocation.sync.token"

  private enum Keys {
    static let auth = "auth_token"
    static let sso = "sso_token"
  }
}

File: android/src/main/java/com/backgroundlocation/sync/TokenStore.kt
package com.backgroundlocation.sync

/**
 * Persists the bearer token used to authenticate bulk-upload requests.
 *
 * An interface (rather than a concrete class directly) so tests can substitute an in-memory
 * fake ([FakeTokenStore]) instead of the real [PrefsTokenStore]. Tokens are pushed in from JS
 * via the module's `setAuthToken`/`setSsoToken`; the sync layer reads them natively at upload
 * time (including while JS is dead), so they live in a plain native prefs file, not the Keychain.
 */
interface TokenStore {
    /** The `x-authorization` bearer token (client/role token). */
    fun getToken(): String?
    fun setToken(token: String?)

    /** The `x-sso-token` value (raw user SSO JWT, sent without a "Bearer" prefix). */
    fun getSsoToken(): String?
    fun setSsoToken(token: String?)
}
